Identity switching and propagation quickstart.

darranl opened the pull request:
This quickstart is to demonstrate how interceptors can be used to facilitate the switching of identities used for EJB invocations and how this capability can also be used to propagate an identity from server to server.

darranl committed:
A new quick start to demonstrate how client and server side interceptors can be used with EJBs to support
identity switching and identity propagation over an existing connection.

darranl committed:
Updated README.md from Sande Gilda.

jaikiran committed:
Further updates to the quickstart optimising how the interceptors are defined.

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+ 
+The quickstart then makes use of two EJBs, `SecuredEJB` and `IntermediateEJB`, to verify that the propagation and identity switching is correct. 
+
+1. The `SecuredEJB` which has three methods: 
```
The numbering doesn't render correctly. Maybe this should be a subheader? For example:

### SecuredEJB

This EJB has three methods:

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+   
+The next two methods are annotated to require that the calling user is authorized for roles 'RoleOne' and 'RoleTwo' respectively.
+
+2. The `IntermediateEJB` contains a single method. Its purpose is to make use of a remote connection and invoke each of the methods on the `SecuredEJB`. A summary is then returned with the outcome of the calls.
```
Like the above, the number doesn't move to 2, so maybe make it a subheader? For example:

### IntermediateEJB

This EJB contains a single method. Its purpose is to ...

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+
+Note on EJB client interceptors
+-----------------------
+AS7/EAP6 allows client side interceptors for EJB invocations. Such interceptors are expected to implement the 'org.jboss.ejb.client.EJBClientInterceptor' interface. User applications can then plug in such interceptors in the 'EJBClientContext' either programatically or
```
Always spell out the product name and put it first ;-)

JBoss Enterprise Applcation Platform 6 and JBoss AS 7 allow client side ...

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+
+If you are reviewing this quickstart with a view to making use of this approach within your own environment it is still
+recommended to try this using a clean installation first and then porting to your own environment. 
+
```
All you need to build this project is Java 6.0 (Java SDK 1.6) or better, Maven 3.0 or better.

The application this project produces is designed to be run on JBoss Enterprise Application Platform 6.1 or JBoss AS 7.2.

I wonder if we should add a Prerequisites section and mention the 6.1/7.2 requirements again. People might miss this here.

Also, I don't understand what  you are trying to say with the sentence : 
"If you are reviewing this quickstart with a view to making use of this approach within your own environment it is still recommended to try this using a clean installation first and then porting to your own environment."
??

darranl replied:
That last line translates to "don't ping me when you have jumped into integrating this into your own environment without first verifying you can get it to work and understand it in a clean environment" ;-)

There are quite a few manual steps to get this to work so understanding it in a controlled environment first if very important.

pmuir replied:
Perhaps just spell that out more clearly. "This quickstart is quite advanced, so we recommend getting the quickstart up and running before trying to use this in your own application".

sgilda commented:
@darranl : I emailed a modified README.md file for you to review. 

Aslo, would  you mind squashing the commits at some point?

pmuir commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+    # on EJB. If these interceptors are to be added at a specific position, other than last, then use the
+    # programmatic API in the application to register it explicitly to the EJBClientContext
+
+    org.jboss.as.quickstarts.ejb_security_interceptors.ClientSecurityInterceptor
```
I possibly wouldn't list out both approaches, but just describe the one you use, and perhaps link to a ref doc or wiki with more info.

pmuir commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,217 @@
+If you have not yet done so, you must [Configure Maven](../README.md#mavenconfiguration) before testing the quickstarts.
+
+
+Add the Application Users
```
Sande, would we normally group all setup steps under one heading?

sgilda replied:
Not really. We group Server Configuration under one heading, then Add the Users would have its own section.

This README needs a little work because the server configuration steps aren't obvious. I've been talking to @darranl on IRC. Also, the server config needs to be more specific about where the XML is placed.

I will help him with this. :-

sgilda replied:
@pmuir: I should mention that this README hasn't been updated with the changes I sent Darran.

sgilda replied:
@pmuir: see this for the example of adding a user: https://github.com/jboss-jdf/jboss-as-quickstart/tree/master/helloworld-jms

pmuir commented:
Basically looks good to me.

sgilda commented:
When I attempt to deploy this application against the latest 7.2, I get this error:

11:41:13,339 INFO  [org.jboss.as.server.deployment](MSC service thread 1-7) JBAS015876: Starting deployment of "jboss-as-ejb-security-interceptors.jar"
11:41:13,425 ERROR [org.jboss.msc.service.fail](MSC service thread 1-3) MSC00001: Failed to start service jboss.deployment.unit."jboss-as-ejb-security-interceptors.jar".PARSE: org.jboss.msc.service.StartException in service jboss.deployment.unit."jboss-as-ejb-security-interceptors.jar".PARSE: JBAS018733: Failed to process phase PARSE of deployment "jboss-as-ejb-security-interceptors.jar"
    at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:127) [jboss-as-server-7.2.0.Alpha1-SNAPSHOT.jar:7.2.0.Alpha1-SNAPSHOT]
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1811) [jboss-msc-1.0.3.GA.jar:1.0.3.GA]
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1746) [jboss-msc-1.0.3.GA.jar:1.0.3.GA]
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886) [rt.jar:1.6.0_33]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908) [rt.jar:1.6.0_33]
    at java.lang.Thread.run(Thread.java:662) [rt.jar:1.6.0_33]
Caused by: java.lang.IllegalStateException: No parser found for urn:container-interceptors:1.0
    at org.jboss.metadata.ejb.parser.spec.ExtendingMetaDataParser.mandatory(ExtendingMetaDataParser.java:50)
    at org.jboss.metadata.ejb.parser.spec.ExtendingMetaDataParser.getParser(ExtendingMetaDataParser.java:45)
    at org.jboss.metadata.ejb.parser.spec.ExtendingMetaDataParser.processElement(ExtendingMetaDataParser.java:80)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossAssemblyDescriptorMetaDataParser.processElement(JBossAssemblyDescriptorMetaDataParser.java:61)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossAssemblyDescriptorMetaDataParser.processElement(JBossAssemblyDescriptorMetaDataParser.java:41)
    at org.jboss.metadata.ejb.parser.spec.AbstractMetaDataParser.processElements(AbstractMetaDataParser.java:69)
    at org.jboss.metadata.ejb.parser.spec.ExtendingMetaDataParser.parse(ExtendingMetaDataParser.java:58)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossEjb3MetaDataParser.parseAssemblyDescriptor(JBossEjb3MetaDataParser.java:119)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossEjb3MetaDataParser.processElement(JBossEjb3MetaDataParser.java:166)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossEjb3MetaDataParser.processElement(JBossEjb3MetaDataParser.java:41)
    at org.jboss.metadata.ejb.parser.spec.AbstractMetaDataParser.processElements(AbstractMetaDataParser.java:69)
    at org.jboss.metadata.ejb.parser.jboss.ejb3.JBossEjb3MetaDataParser.parse(JBossEjb3MetaDataParser.java:113)
    at org.jboss.as.ejb3.deployment.processors.EjbJarParsingDeploymentUnitProcessor.parseJBossEjb3Xml(EjbJarParsingDeploymentUnitProcessor.java:276)
    at org.jboss.as.ejb3.deployment.processors.EjbJarParsingDeploymentUnitProcessor.deploy(EjbJarParsingDeploymentUnitProcessor.java:119)
    at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:120) [jboss-as-server-7.2.0.Alpha1-SNAPSHOT.jar:7.2.0.Alpha1-SNAPSHOT]
    ... 5 more

11:41:13,433 ERROR [org.jboss.as.server](management-handler-thread - 1) JBAS015870: Deploy of deployment "jboss-as-ejb-security-interceptors.jar" was rolled back with the following failure message: 
{"JBAS014671: Failed services" => {"jboss.deployment.unit.\"jboss-as-ejb-security-interceptors.jar\".PARSE" => "org.jboss.msc.service.StartException in service jboss.deployment.unit.\"jboss-as-ejb-security-interceptors.jar\".PARSE: JBAS018733: Failed to process phase PARSE of deployment \"jboss-as-ejb-security-interceptors.jar\"
    Caused by: java.lang.IllegalStateException: No parser found for urn:container-interceptors:1.0"}}
11:41:13,441 INFO  [org.jboss.as.server.deployment](MSC service thread 1-4) JBAS015877: Stopped deployment jboss-as-ejb-security-interceptors.jar in 8ms
^X^C16:12:20,481 INFO  [org.apache.catalina.core](MSC service thread 1-7) JBWEB001079: Container org.apache.catalina.core.ContainerBase.[jboss.web].[default-host].[/] has not been started

sgilda commented:
I downloaded 7.2 from here, https://ci.jboss.org/hudson/job/JBoss-AS-7.x-latest/lastBuild/, and it works great! Apparently I didn't have the correct version of the server.

darranl committed:
Subsequent README update from Sande Gilda.

darranl committed:
Changes to README to have a section for defining the configuration using the CLI and a section for manually editing the standalone.xml configuration.

Also a change to a new security domain called 'quickstart-domain' as using the CLI means we need a new domain instead of modifying an existing domain.

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,312 @@
+[standalone@localhost:9999 /] ./subsystem=security/security-domain=quickstart-domain/authentication=classic:add
+[standalone@localhost:9999 /] ./subsystem=security/security-domain=quickstart-domain/authentication=classic/login-module=DelegationLoginModule:add(code=org.jboss.as.quickstarts.ejb_security_interceptors.DelegationLoginModule,flag=optional,module-options={password-stacking=useFirstPass})    
+[standalone@localhost:9999 /] ./subsystem=security/security-domain=quickstart-domain/authentication=classic/login-module=Remoting:add(code=Remoting,flag=optional,module-options={password-stacking=useFirstPass})
+[standalone@localhost:9999 /] ./subsystem=security/security-domain=quickstart-domain/authentication=classic/login-module=RealmDirect:add(code=RealmDirect,flag=required,module-options={password-stacking=useFirstPass})
```
All of these CLI commands need to be indented 8 spaces so they appear as commands for easier readability. Same for the ones below.

sgilda commented on ejb-security-interceptors/README.md:
```diff
@@ -0,0 +1,312 @@
+| ConnectionUser| ConnectionPassword1!| User |
+| AppUserOne | (you can choose any password) | User, RoleOne |
+| AppUserTwo | (you can choose any password) | User, RoleTwo |
+| AppUserThree | (you can choose any password) | User, RoleOne, RoleTwo |
```
Were you going to provide passwords for the other users so they  won't have to come up with them? Maybe:
User1Pwd!
User2Pwd!
User3Pwd!

Or something like that?

sgilda commented:
Darran: This looks great except for a few minor typos. Is it possible to squash the commits?

@pmuir: Do you need to do a final code review?

darranl commented:
That should now hopefully be the final update to this pull request.

I have now specified some passwords to use for the additional users and added a note to say that we don't actually use those passwords within the quickstart - I have also updated the hashes for the users to match as I could not remember what passwords I had set!! ;-)

Regarding squashing this is squashed as low as I can get it - squashing any further will mean the changes will no longer be attributed to the correct author.

sgilda commented:
Looks great! Merged upstream.
